packages/@lwc/engine-core/src/framework/decorators/register.ts

```diff
@@ -58,15 +57,26 @@ interface RegisterDecoratorMeta {
     readonly fields?: string[];
 }
 
+function getClassDescriptorType(descriptor: PropertyDescriptor): string {
+    if (isFunction(descriptor.value)) {
+        return 'method';
+    } else if (isFunction(descriptor.set) || isFunction(descriptor.get)) {
+        return 'accessor';
+    } else {
+        return 'field';
+    }
+}
+
 function validateObservedField(
     Ctor: ComponentConstructor,
     fieldName: string,
     descriptor: PropertyDescriptor | undefined
 ) {
-    if (process.env.NODE_ENV !== 'production') {
-        if (!isUndefined(descriptor)) {
-            assert.fail(`Compiler Error: Invalid field ${fieldName} declaration.`);
-        }
+    if (!isUndefined(descriptor)) {
```

Comment:
I'm always a little nervous about the function dependency graph for development mode only functionality. It is too easy to mess up, and use one of those methods from another file that doesn't do the wrapping. In the past I used a single line to throw if we are running in prod mode, instead of just wrapping, e.g.: 

```
  if (process.env.NODE_ENV === 'production') {
    // this function should never leak to prod
    throw new ReferenceError();
  }
```

Reply:
About dev-only validation, we had in the past a build step on our CI that would validate that those restrictions are only enabled in DEV mode by running tests PROD mode. 

It's not something that we ported to the Karma integration tests (to my surprise). If we were to able this again, would it remove some of your concerns with this?